cip/vendor-extensions/neo4j/CIP2016-12-14-Neo4j-indexes.adoc

```diff
@@ -0,0 +1,138 @@
+= CIP2016-12-16 - Neo4j Indexes
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Author:* Mats Rydberg <mats@neotechnology.com>
+
+[abstract]
+.Abstract
+--
+This CIP details Neo4j's indexing extension to Cypher, which is based on the standardised constraints syntax.
+--
+
+toc::[]
+
+== Background
+
+In Neo4j, indexes are formed using label and property combinations.
+This enables queries that reference these label/property combinations to use the index for faster lookup with reduced cardinality overhead.
+
+== Proposal
+
+Indexes in Neo4j are able to index _labeled nodes_ only.
+These nodes are kept in a separate, persisted data structure which allows lookups based on providing values for the specified indexed properties.
+
+=== Syntax
+
+The index syntax is based on the constraint syntax, and is detailed below:
+
+.Grammar definition for Neo4j index syntax.
+[source, ebnf]
+----
+index command = create-index | drop-index ;
+create-index  = "CREATE", "INDEX", [ index-name ], "FOR", index-pattern, "ON", index-key ;
+index-pattern = node-pattern
+index-name    = symbolic-name
+index-key     = property-expression { ",", property-expression } ;
+drop-index    = "DROP", "INDEX", index-name ;
+----
+
+The `index-key` expression defines the key for the index, and consist of one or more property expressions, which refer to the entity defined in the pattern.
+
+==== Index names
+
+Just like constraints, indexes have names.
+Neo4j does not support user-defined names, but the index will be assigned a system-generated name.
+
+==== Removing indexes
+
+An index is removed by referring to its name.
```

Comment:
The system will generate a name. Only the 'user-defined' part is optional; they should always have a name.